openff/toolkit/typing/engines/smirnoff/parameters.py

```diff
@@ -3783,15 +3805,80 @@ def scale15(self, attrs, new_scale15):
+                    "See https://openforcefield.github.io/standards/standards/smirnoff/#electrostatics"
+                )
+        if kwargs.get("version") == 0.3:
+            logger.info(
+                "Attempting to up-convert Electrostatics section from 0.3 to 0.4"
+            )
+            # Default value in 0.3 is "PME", so we have to handle these cases identically
+            if kwargs.get("method") in ["PME", None]:
+                kwargs["periodic_potential"] = "Ewald3D-ConductingBoundary"
+                kwargs["nonperiodic_potential"] = "Coulomb"
+                kwargs["exception_potential"] = "Coulomb"
+                kwargs["version"] = 0.4
+                kwargs.pop("method", None)
+                logger.info(
+                    'Successfully up-converted Electrostatics section from 0.3 to 0.4. `method="PME"` '
+                    'is now split into `periodic_potential="Ewald3D-ConductingBoundary"`, '
+                    '`nonperiodic_potential="Coulomb"`, and `exception_potential="Coulomb"`.'
+                )
+            elif kwargs["method"] == "Coulomb":
+                kwargs["periodic_potential"] = "Coulomb"
+                kwargs["nonperiodic_potential"] = "Coulomb"
+                kwargs["exception_potential"] = "Coulomb"
+                kwargs["version"] = 0.4
+                kwargs.pop("method", None)
+                logger.info(
+                    'Successfully up-converted Electrostatics section from 0.3 to 0.4. `method="Coulomb"` '
+                    'is now split into `periodic_potential="Coulob"`, '
+                    '`nonperiodic_potential="Coulomb"`, and `exception_potential="Coulomb"`.'
+                )
+            elif kwargs["method"] == "reaction-field":
+                kwargs["periodic_potential"] = self._DEFAULT_REACTION_FIELD_EXPRESSION
+                kwargs["nonperiodic_potential"] = "Coulomb"
+                kwargs["exception_potential"] = "Coulomb"
+                kwargs["version"] = 0.4
+                kwargs.pop("method", None)
+                logger.info(
+                    'Successfully up-converted Electrostatics section from 0.3 to 0.4. `method="Coulomb"` '
+                    f'is now split into `periodic_potential="{self._DEFAULT_REACTION_FIELD_EXPRESSION}"` '
+                    '`nonperiodic_potential="Coulomb"`, and `exception_potential="Coulomb"`.'
+                )
+            else:
+                # This clause shouldn't be accessible; the only previously-allowed values
+                # (`converter=_allow_only(["Coulomb", "PME", "reaction-field"])`)
+                # should all be handled by the above up-conversion
+                logger.warning(
+                    "Failed to up-convert Electrostatics section from 0.3 to 0.4. Did not know "
+                    f"how to up-convert `method={kwargs['method']}`. Continuing to use version 0.3."
+                )
```

Comment:
This seems dangerous and untested/able - I'm thinking it should be a NotImplementedError.

Reply:
Sure, this was only a fallback because I thought raising a new exception was not going to be favored.